network: add a UDP test

mardy opened the pull request:
I actually wrote this to debug an issue I'm having with my Wii, where TCP connections work, but UDP do not. On Dolphin this works.

I would appreciate if someone (I'm especially hopeful on @DacoTaco :-) ) could give this a try on a console, and report if it's working.

To test it, launch the program, then from a PC run the following command:

    nc -u <wii-ip-address> 45679
    # Then type something, and press ENTER

DacoTaco commented:
fuck, i want to test this but i wii has a shit ton of issues connecting to my wifi recently haha.

ill see if i can make it work :)

mardy commented:
> ill see if i can make it work :)

One thing I noticed is that net_poll() correctly returns that that there is data (when I send something via `nc`), but then `net_recvfrom()` returns a -95 (which seems to be EOPNOTSUPP).

DacoTaco commented:
> > ill see if i can make it work :)
> 
> One thing I noticed is that net_poll() correctly returns that there is data (when I send something via `nc`), but then `net_recvfrom()` returns a -95 (which seems to be EOPNOTSUPP).

should the app be doing something when i press enter? it doesn't seem to be responding to my input into netcat :/

mardy commented:
> should the app be doing something when i press enter? it doesn't seem to be responding to my input into netcat :/

Yes, it should print "Echo: <what you sent>". It does that on Dolphin, but not on the Wii.

> also said own address to be 0.0.0.0, which it isn't :p

True, that seems an issue with `net_getsockname()`: it returns the port, but not the IP.

DacoTaco commented:
hmmm.
now do we merge this now, with it not working on hardware or first find the UDP issue...

i think finding the UDP issue is best looked at first

mardy commented:
> hmmm. now do we merge this now, with it not working on hardware or first find the UDP issue...
> 
> i think finding the UDP issue is best looked at first

Fully agree. For now I mainly shared it to see if it was an issue with my Wii only. I'll continue investigating, then.

By the way, the 0.0.0.0 address for the socket name is actually expected, on Linux `getsockname()` does the same, if it was given a `INADDR_ANY` value to net_bind().

mardy commented:
OK, it looks like it's working now. The problem was the MSG_DONTWAIT flag; for some reason, specifying it makes the net_recvfrom call fail. I added a comment about it in the source code, since in many cases it can be tempting to add it.

I also added a few lines of help, that show the console IP address and the `nc` command to run in order to test the example.

DacoTaco commented:
confirmed working and a bit more obvious to the person running it what it does/can be tested :)
